content/blog/2025-12-15-brilir.md

```diff
@@ -0,0 +1,104 @@
+
+The implementation can be broken down into four major components:
+
+- The Bril dialect definition
+- The Bril -> MLIR converter
+- The MLIR -> Bril converter
+- The Bril -> `std` dialect conversion pass
+
+## The Bril Dialect
+
+The Bril MLIR dialect is expressed using MLIR's [Operation Definition Specification](https://mlir.llvm.org/docs/DefiningDialects/Operations/) (ODS), which is MLIR's DSL for declaratively defining operations, types, and other dialect-related objects. ODS is the recommended approach for defining MLIR dialects, as it significantly reduces boilerplate C++ code and integrates cleanly with CMake via the TableGen tool, which generates the corresponding C++ sources and headers.
+
+In this project, ODS is used to define Bril operations, types, and passes. I chose to reuse MLIR's standard integer type to represent Bril's `int` and `bool` types. For the memory extension, I defined a dedicated opaque pointer type to represent Bril pointers.
+
+Each Bril operation corresponds closely to an operation in the dialect. Arithmetic operations are straightforward. More interesting cases include operations such as `const`, which takes an attribute operand (representing a constant value in MLIR); the function operation `func`, which contains a region of basic blocks; the `call` operation, which references an MLIR symbol corresponding to a `func`; and control-flow operations such as `br`, which specify successor blocks and pass block arguments (since MLIR models SSA via block arguments).
+
+Finally, I defined two passes: `convert-bril-to-std`, a dialect conversion pass that lowers Bril into standard MLIR dialects and `llvm` as a first step toward LLVM IR, and `rename-main-function`, a somewhat hacky pass used to rename the Bril entry-point function for testing purposes (discussed further below).
+
+## Bril to MLIR Converter
+
+Bril's JSON-based representation makes parsing and conversion relatively straightforward, as I can directly rely on a [JSON library](https://github.com/nlohmann/json) to read the input program.
+
+Since MLIR operates in SSA form, the converter expects the input Bril program to already be in SSA form, specifically using the `get` / `set` representation. This maps fairly naturally to MLIR's block-argument-based SSA representation, aside from several corner cases discussed below.
+
+The converter first constructs basic blocks for each function and then performs SSA conversion. To translate `get` / `set` into block arguments, I collect all `get` instructions in a block and treat them as that block's arguments. In MLIR, block arguments are passed through control-flow operations, so I also collect each `set` instruction in a block and use those values to populate the operands of control-flow terminators, except for blocks terminated by a return.
```

Comment:
Any chance this runs into issues if the `get` and `set` are not at the top and bottom of the block, respectively? I think the answer is no (because we're in SSA, and because `set`s are SSU, it's always equivalent to move them up or down), but just wanted to check.